gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp

```diff
@@ -698,114 +698,160 @@ OGRErr OGRFlatGeobufLayer::parseFeature(OGRFeature *poFeature, OGRGeometry **ogr
     return OGRERR_NONE;
 }
 
-OGRPoint *OGRFlatGeobufLayer::readPoint(const Feature *feature, uint32_t offset)
+OGRPoint *OGRFlatGeobufLayer::readPoint(const Feature *feature, const flatbuffers::Vector<double> *pXy, uint32_t offset)
 {
-    auto aXy = feature->xy()->data();
+    auto aXy = pXy->data();
     if (m_hasZ) {
-        auto aZ = feature->z()->data();
+        auto pZ = feature->z();
+        if (pZ == nullptr)
+            CPLErrorInvalidPointer();
+        auto aZ = pZ->data();
         if (m_hasM) {
-            auto aM = feature->m()->data();
+            auto pM = feature->m();
+            if (pM == nullptr)
+                CPLErrorInvalidPointer();
+            auto aM = pM->data();
             return new OGRPoint { aXy[offset + 0], aXy[offset + 1], aZ[offset], aM[offset] };
         } else {
             return new OGRPoint { aXy[offset + 0], aXy[offset + 1], aZ[offset] };
         }
     } else if (m_hasM) {
-        auto aM = feature->m()->data();
+        auto pM = feature->m();
+        if (pM == nullptr)
+            CPLErrorInvalidPointer();
+        auto aM = pM->data();
         return new OGRPoint { aXy[offset + 0], aXy[offset + 1], 0.0, aM[offset] };
     } else {
         return new OGRPoint { aXy[offset + 0], aXy[offset + 1] };
     }
 }
 
-OGRMultiPoint *OGRFlatGeobufLayer::readMultiPoint(const Feature *feature, uint32_t len)
+OGRMultiPoint *OGRFlatGeobufLayer::readMultiPoint(const Feature *feature, const flatbuffers::Vector<double> *pXy, uint32_t len)
 {
     auto mp = new OGRMultiPoint();
     for (uint32_t i = 0; i < len; i = i + 2)
-        mp->addGeometryDirectly(readPoint(feature, i));
+        mp->addGeometryDirectly(readPoint(feature, pXy, i));
```

Comment:
in the case of a Z,M,ZM geometry, isn't there an issue with the index of the Z and M arrays ? It looks to me that it will be 2 times its expected value. And there are likely missing checks here also on the offset regarding the size of the xy, z and m arrays.

Reply:
in the unit tests, I don't see any testing of MULTIPOINT Z/M/ZM

Reply:
the return value of readPoint() isn't also checked against nullptr